packages/aws-rfdk/lib/core/lambdas/nodejs/version-provider/version-provider.ts

```diff
@@ -0,0 +1,323 @@
+
+    const requestedVersion = this.parseVersionString( versionString );
+
+    // Based on the requested version, fetches the latest patch and its installer file paths.
+    return this.getRequestedUriVersion(
+      requestedVersion,
+      productSection.versions,
+      platform,
+      product,
+    );
+  }
+
+  /**
+   * This method returns the latest version for specified platform.
+   *
+   * @param platform
+   * @param indexedVersionInfo
+   */
+  private getLatestVersion(platform: string, indexedVersionInfo: any): string {
+    const latestSection = indexedVersionInfo.latest;
+    if (!latestSection) {
+      throw new Error('Information about latest version can not be found');
+    }
+
+    const latestVersion = latestSection[platform];
+    if (!latestVersion) {
+      throw new Error(`Information about latest version for platform ${platform} can not be found`);
+    }
+
+    return latestVersion;
+  }
+
+  /**
+   * This method looks for the requested version (partial or complete) in the
+   * indexed version information. Based on the input, it iterates through all
+   * four numbers in the version string and compares the requested version
+   * with the indexed info.
+   * If any of the requested version number is missing, it fetches the latest
+   * (highest) available version for it.
+   *
+   * @param requestedVersion
+   * @param indexedVersionInfo
+   */
+  private getRequestedUriVersion(
+    requestedVersion: RegExpExecArray | null,
+    indexedVersionInfo: any,
+    platform: Platform,
+    product: Product,
+  ): IVersionedUris | undefined {
+
+    let versionMap = indexedVersionInfo;
+    const versionArray: string[] = [];
+
+    // iterate through all 4 major, minor, release and patch numbers,
+    // and get the matching version from the indexed version map.
+    for (let versionIndex = 0; versionIndex < 4; versionIndex++) {
+      let version: string;
+      if (null === requestedVersion
+        || null === requestedVersion[versionIndex + 1]
+        || typeof requestedVersion[versionIndex + 1] === 'undefined') {
+
+        // version is not provided, get the max version.
+        const numberValues: number[] = (Object.keys(versionMap)).map((val: string) => {
+          return parseInt(val, 10);
+        });
+        version = (Math.max(...numberValues)).toString();
+
+      } else {
+        version = requestedVersion[versionIndex + 1];
+      }
+      versionArray[versionIndex] = version;
+      versionMap = versionMap[version];
+    }
+
+    let uriIndex: IDeadlineInstallerUris | IDeadlineDockerRecipeUris | undefined;
+    if ((platform in versionMap)) {
+      if (product == Product.deadline) {
+        uriIndex = {
+          bundle: versionMap[platform].bundle,
+          clientInstaller: versionMap[platform].clientInstaller,
+          repositoryInstaller: versionMap[platform].repositoryInstaller,
+          certificateInstaller: versionMap[platform].certificateInstaller,
```

Comment:
I don't think this will work for the Mac section, it does not have repository or client installers, only the bundle. Our interface does not have repository and client as optional. We should write a unit test that covers this case, confirm it fails, and then update the interface to use optionals here.